### PR TITLE
PIM-4666: Sakafix

### PR DESCRIPTION
**Pull request by sumbobyboys**
| Q                  | A
| ------------------ | ---
| Migration script?  | N
| Behats             | ?
| Specs & Static     | N
| Changelog updated  | N

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php [hunk 1]
```diff
@@ -70,6 +70,7 @@ public function normalize($product, $format = null, array $context = [])
         $updated = null !== $newestLog ? $this->versionNormalizer->normalize($newestLog, 'internal_api') : null;
 
         $normalizedProduct['meta'] = [
+            'form'              => 'product-edit-form',
```
Comment (jmleroux) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php]: could we not use a constant ?

File: features/Context/DataGridContext.php [hunk 2]
```diff
@@ -867,7 +867,7 @@ public function iPressMassDeleteButton()
     public function iPressSequentialEditButton()
     {
         $this->getCurrentPage()->sequentialEdit();
-        $this->wait();
+        $this->wait(20000);
```
Comment (sumbobyboys) [inline features/Context/DataGridContext.php]: I will maybe drop this one

File: features/Context/AssertionContext.php [hunk 3]
```diff
@@ -172,9 +172,8 @@ public function iShouldNotSeeTheFields($fields)
                 if ($this->getCurrentPage()->findField($field)) {
                     throw $this->createExpectationException(sprintf('Not expecting to see field "%s"', $field));
                 }
+            } catch (ElementNotFoundException $e) {
```
Comment (phaseinducer) [inline features/Context/AssertionContext.php]: two catch and nothing inside ?

Comment (sumbobyboys) [inline features/Context/AssertionContext.php]: Yup, if nothing is catch, it means that we didn't found the element. And if something was found, we throw an expectation exception;

File: src/Pim/Bundle/EnrichBundle/DependencyInjection/Compiler/RegisterGenericProvidersPass.php [hunk 4]
```diff
@@ -7,32 +7,37 @@
...
 
+    /** @var string */
+    protected $providerType;
+
     /**
      * @param ReferenceFactory $factory
+     * @param string           $providerType
      */
-    public function __construct(ReferenceFactory $factory)
+    public function __construct(ReferenceFactory $factory, $providerType)
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/DependencyInjection/Compiler/RegisterGenericProvidersPass.php]: DIception :trollface:

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php [hunk 5]
```diff
@@ -70,6 +77,7 @@ public function normalize($product, $format = null, array $context = [])
         $updated = null !== $newestLog ? $this->versionNormalizer->normalize($newestLog, 'internal_api') : null;
 
         $normalizedProduct['meta'] = [
+            'form'              => $this->formProvider->getForm($product),
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php]: what's a form here ?

File: src/Pim/Bundle/EnrichBundle/Provider/Form/FormProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Pim\Bundle\EnrichBundle\Provider\Form;
+
+/**
+ * Form provider interface
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Provider/Form/FormProviderInterface.php]: explain what it is please, I have no idea...

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php (hunk 5 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php]: A string defining which javascript form you should load for the given product :)

File: features/product/validation/remove_validation_error_on_correction.feature [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  Scenario: Validate that the validation error is removed on error correction
+    Given I change the "Cost" to "-10 USD"
+    And I save the product
+    Then I should see validation tooltip "This value should be 0 or more."
+    And there should be 1 error in the "Other" tab
+    Then I change the "Cost" to "10 USD"
+    And I save the product
+    Then I should not see validation tooltip "This value should be 0 or more."
+    And there should be 0 error in the "Other" tab
+
```
Comment (phaseinducer) [inline features/product/validation/remove_validation_error_on_correction.feature]: :+1:

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php (hunk 5 above)
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php]: edit or view ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/manager/product-manager.js [hunk 8]
```diff
@@ -57,16 +57,16 @@ define([
...
                             product.values[attributeCode] = [];
                         }
-                    }, this));
+                    }.bind(this));
 
                     return product.values;
                 });
             },
-            generateMissing: function (product) {
+            _generateMissing: function (product) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/manager/product-manager.js]: doGenerateMissing ? Not a huge fan of this "_" convention : /

Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/manager/product-manager.js]: We need to document that, it's very confusing

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js [hunk 9]
```diff
@@ -486,9 +486,9 @@ define(
                     url: this.sortingUrl,
                     type: 'PUT',
                     data: JSON.stringify(sorting)
-                }).done(_.bind(function () {
+                }).done(function () {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: `.then()` ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js [hunk 10]
```diff
@@ -44,13 +44,13 @@ define(
...
                     this.render();
-                }, this));
+                }.bind(this));
             },
             renderInput: function (context) {
                 return this.fieldTemplate(context);
             },
             postRender: function () {
-                this.getChoiceUrl().done(_.bind(function (choiceUrl) {
+                this.getChoiceUrl().done(function (choiceUrl) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js [hunk 11]
```diff
@@ -36,7 +36,7 @@ define(
                 if (!SecurityContext.isGranted('pim_enrich_attribute_edit')) {
                     return;
                 }
-                createOption(this.attribute).done(_.bind(function (option) {
+                createOption(this.attribute).done(function (option) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/simple-select-field.js [hunk 12]
```diff
@@ -37,19 +37,19 @@ define(
...
                     this.render();
-                }, this));
+                }.bind(this));
             },
             renderInput: function (context) {
                 return this.fieldTemplate(context);
             },
             postRender: function () {
-                this.getChoiceUrl().done(_.bind(function (choiceUrl) {
+                this.getChoiceUrl().done(function (choiceUrl) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/simple-select-field.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js [hunk 13]
```diff
@@ -22,21 +22,40 @@ define(
...
                     this.configured = true;
-                }, this));
+                }.bind(this));
             },
+
+            /**
+             * Add a child extension to this extension
+             *
+             * @param {string} code      Extension's code
+             * @param {object} extension Backbone module of the extension
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js]: Object (capitalize)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js [hunk 14]
```diff
@@ -46,14 +65,47 @@ define(
 
                 this.extensions[code] = extension;
             },
+
+            /**
+             * Get a child extension (the first extension matching the given code or ends with the given code)
+             *
+             * @param {string} code
+             *
+             * @return {object}
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js]: same

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js [hunk 15]
```diff
@@ -46,14 +65,47 @@ define(
...
+                    var expectedPosition = extension.code.length - code.length;
+
+                    return expectedPosition >= 0 && expectedPosition === extension.code.indexOf(code, expectedPosition);
+                })];
+            },
+
+            /**
+             * Set the parent of this extension
+             *
+             * @param {object} parent
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js]: same

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 16]
```diff
@@ -120,7 +120,7 @@ define(
                 $.when(
                     this.loadAssociationTypes(),
                     FetcherRegistry.getFetcher('attribute').getIdentifierField()
-                ).done(_.bind(function (associationTypes, identifierAttribute) {
+                ).done(function (associationTypes, identifierAttribute) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js [hunk 17]
```diff
@@ -83,44 +83,44 @@ define(
                 }
 
                 this.rendering = true;
-                this.getConfig().done(_.bind(function () {
+                this.getConfig().done(function () {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js [hunk 18]
```diff
@@ -198,8 +198,8 @@ define(
                 Dialog.confirm(
                     _.__('pim_enrich.confirmation.delete.product_attribute'),
                     _.__('pim_enrich.confirmation.delete_item'),
-                    _.bind(function () {
-                        FetcherRegistry.getFetcher('attribute').fetch(attributeCode).done(_.bind(function (attribute) {
+                    function () {
+                        FetcherRegistry.getFetcher('attribute').fetch(attributeCode).done(function (attribute) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js [hunk 19]
```diff
@@ -247,7 +249,7 @@ define(
             },
             showAttribute: function (event) {
                 AttributeGroupManager.getAttributeGroupsForProduct(this.getFormData())
-                    .done(_.bind(function (attributeGroups) {
+                    .done(function (attributeGroups) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes.js]: .then()

File: src/Pim/Bundle/EnrichBundle/Resources/views/Product/edit.html.twig [hunk 20]
```diff
@@ -58,16 +57,17 @@
...
-                                        FormBuilder.build('pim/product-edit-form')
-                                            .done(function (form) {
-                                                ProductManager.get('{{ productId }}').done(function (product) {
-                                                    document.title = document.title.replace('%product.sku%', _.escape(product.meta.label[UserContext.get('catalogLocale')]));
-                                                    form.setData(product);
+                                        ProductManager.get('{{ productId }}')
+                                            .then(function (product) {
+                                                FormBuilder.build(product.meta.form)
+                                                    .then(function (form) {
+                                                        document.title = document.title.replace('%product.sku%', _.escape(product.meta.label[UserContext.get('catalogLocale')]));
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/views/Product/edit.html.twig]: We don't have to wait for that the form is build to do this

File: src/Pim/Component/Catalog/Comparator/Attribute/PricesComparator.php [hunk 21]
```diff
@@ -42,14 +42,19 @@ public function compare($data, $originals)
```
Comment (phaseinducer) [inline src/Pim/Component/Catalog/Comparator/Attribute/PricesComparator.php]: Can you align the lines above please ?

File: src/Pim/Component/Catalog/Comparator/Attribute/PricesComparator.php [hunk 22]
```diff
@@ -42,14 +42,19 @@ public function compare($data, $originals)
...
 
         $dataPrices = [];
         foreach ($data['data'] as $price) {
-            $dataPrices[$price['currency']] = number_format($price['data'], 4);
+            if (null !== $price['data']) {
+                $dataPrices[$price['currency']] = number_format($price['data'], 4);
+            }
         }
 
+
```
Comment (phaseinducer) [inline src/Pim/Component/Catalog/Comparator/Attribute/PricesComparator.php]: remove blank line

**Comment by phaseinducer** [comment]
Your specs are out of date.

I didn't wrote it everywhere but :
- Object in jsdoc is {Object} (with a capital)
- replace `.done()` by `.then()` if you have enough time

Otherwise, awesome PR =)

File: src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php (hunk 5 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Normalizer/ProductNormalizer.php]: yep `pim-product-edit-form` or `pim-product-view-form` (or anything else that an integrator could do :))

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/attribute-option/create.js [hunk 23]
```diff
@@ -82,9 +82,9 @@ define(
                                     _.__('pim_enrich.form.attribute_option.flash.error_creating_option')
                                 );
                             }
-                        }, this));
-                    }, this));
-                }, this));
+                        }.bind(this));
+                    }.bind(this));
+                }.bind(this));
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/attribute-option/create.js]: \o/

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js (hunk 9 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: `.done()` is correct here IMHO, as we don't alter the original promise

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/edit-form.js [hunk 24]
```diff
@@ -37,6 +37,10 @@ define(
                 mediator.clear('pim_enrich:form');
                 Backbone.Router.prototype.once('route', this.unbindEvents);
 
+                this.onExtensions('save-buttons:register-button', function (button) {
+                    this.getExtension('save-buttons').trigger('save-buttons:add-button', button);
+                });
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/edit-form.js]: I think it misses a `.bind(this)` on the first anonymous fonction

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js (hunk 11 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/multi-select-field.js]: I think `.done()` is ok here, we don't alter the promise

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/simple-select-field.js (hunk 12 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/simple-select-field.js]: same as above, to me `done()` is ok here

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js [hunk 25]
```diff
@@ -22,21 +22,40 @@ define(
...
+             */
             initialize: function () {
                 this.extensions   = {};
                 this.zones        = {};
                 this.targetZone   = '';
                 this.configured   = false;
             },
+
+            /**
+             * Configure the extension and it's child extensions
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js]: its

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js [hunk 26]
```diff
@@ -81,12 +141,28 @@ define(
 
                 return this;
             },
+
+            /**
+             * Get the form raw data (vanila javascript object)
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form.js]: vanilla

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes/copy.js [hunk 27]
```diff
@@ -117,6 +118,10 @@ define(
                         this.scope
                     );
 
+                    if (_.isUndefined(valueToCopy)) {
+                        valueToCopy = AttributeManager.generateValue(field.attribute, this.locale, this.scope)
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/attributes/copy.js]: Missing semicolon at the end of file

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js (hunk 9 above)
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/pim-attributeoptionview.js]: Just .bind refactor, I will not touch this code more now

**Comment by phaseinducer** [comment]
GTM

**Comment by nidup** [comment]
GTM

**Pull request merged**